File: frontend/src/types/content.ts
export type ContentLanguage = 'en' | 'te'
export type ContentStatus = 'draft' | 'published' | 'unpublished' | 'scheduled'
export type NewsType = 'featured' | 'latest' | 'trending' | 'more'

export const NEWS_TYPES: { value: NewsType; label: string }[] = [
  { value: 'featured', label: 'Featured' },
  { value: 'latest', label: 'Latest' },
  { value: 'trending', label: 'Trending' },
  { value: 'more', label: 'More' },
]

export interface Category {
  id: string
  name: string
  slug: string
  description: string | null
  is_active: boolean
  created_at: string
}

export interface TagItem {
  id: string
  name: string
  slug: string
  created_at: string
}

export interface NewsItem {
  id: string
  title: string
  slug: string
  language: ContentLanguage
  short_description: string | null
  content: string
  category_id: string | null
  category_name: string | null
  author_id: string | null
  author_name: string | null
  status: ContentStatus
  news_type: NewsType
  is_featured: boolean
  is_breaking: boolean
  image_url: string | null
  seo_title: string | null
  seo_description: string | null
  seo_keywords: string | null
  published_at: string | null
  view_count: number
  created_at: string
  updated_at: string
  tags: TagItem[]
}

export interface PaginatedNews {
  items: NewsItem[]
  total: number
  page: number
  page_size: number
}

export interface NewsPayload {
  title: string
  slug: string
  language: ContentLanguage
  short_description?: string | null
  content?: string
  category_id?: string | null
  tag_ids?: string[]
  status?: ContentStatus
  news_type?: NewsType
  is_featured?: boolean
  is_breaking?: boolean
  image_url?: string | null
  seo_title?: string | null
  seo_description?: string | null
  seo_keywords?: string | null
  published_at?: string | null
}

export interface BlogItem {
  id: string
  title: string
  slug: string
  language: ContentLanguage
  short_description: string | null
  content: string
  category_id: string | null
  category_name: string | null
  author_id: string | null
  author_name: string | null
  status: ContentStatus
  image_url: string | null
  seo_title: string | null
  seo_description: string | null
  seo_keywords: string | null
  published_at: string | null
  view_count: number
  created_at: string
  updated_at: string
  tags: TagItem[]
}

export interface PaginatedBlogs {
  items: BlogItem[]
  total: number
  page: number
  page_size: number
}

export interface BlogPayload {
  title: string
  slug: string
  language: ContentLanguage
  short_description?: string | null
  content?: string
  category_id?: string | null
  tag_ids?: string[]
  status?: ContentStatus
  image_url?: string | null
  seo_title?: string | null
  seo_description?: string | null
  seo_keywords?: string | null
  published_at?: string | null
}

File: backend/app/models/content.py
import enum
import uuid
from datetime import datetime, timezone

from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class ContentStatus(str, enum.Enum):
    draft = "draft"
    published = "published"
    unpublished = "unpublished"
    scheduled = "scheduled"


class ContentLanguage(str, enum.Enum):
    en = "en"
    te = "te"


class NewsType(str, enum.Enum):
    featured = "featured"
    latest = "latest"
    trending = "trending"
    more = "more"


class Category(Base):
    __tablename__ = "categories"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    slug: Mapped[str] = mapped_column(String(160), unique=True, index=True, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )


class Tag(Base):
    __tablename__ = "tags"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    slug: Mapped[str] = mapped_column(String(160), unique=True, index=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )


class News(Base):
    __tablename__ = "news"
    __table_args__ = (UniqueConstraint("language", "slug", name="uq_news_language_slug"),)

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    slug: Mapped[str] = mapped_column(String(320), index=True, nullable=False)
    language: Mapped[ContentLanguage] = mapped_column(
        Enum(ContentLanguage, native_enum=False), index=True, nullable=False
    )
    short_description: Mapped[str | None] = mapped_column(Text, nullable=True)
    content: Mapped[str] = mapped_column(Text, default="", nullable=False)
    category_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("categories.id"), nullable=True)
    author_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("users.id"), nullable=True)
    status: Mapped[ContentStatus] = mapped_column(
        Enum(ContentStatus, native_enum=False), default=ContentStatus.draft, index=True
    )
    news_type: Mapped[NewsType] = mapped_column(
        Enum(NewsType, native_enum=False),
        default=NewsType.latest,
        index=True,
        nullable=False,
    )
    is_featured: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    is_breaking: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    is_local: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    image_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    seo_title: Mapped[str | None] = mapped_column(String(300), nullable=True)
    seo_description: Mapped[str | None] = mapped_column(Text, nullable=True)
    seo_keywords: Mapped[str | None] = mapped_column(String(500), nullable=True)
    published_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    view_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    category = relationship("Category")
    author = relationship("User")
    tags = relationship("Tag", secondary="news_tags", lazy="selectin")


class NewsTag(Base):
    __tablename__ = "news_tags"

    news_id: Mapped[str] = mapped_column(String(36), ForeignKey("news.id"), primary_key=True)
    tag_id: Mapped[str] = mapped_column(String(36), ForeignKey("tags.id"), primary_key=True)


class Blog(Base):
    __tablename__ = "blogs"
    __table_args__ = (UniqueConstraint("language", "slug", name="uq_blogs_language_slug"),)

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    slug: Mapped[str] = mapped_column(String(320), index=True, nullable=False)
    language: Mapped[ContentLanguage] = mapped_column(
        Enum(ContentLanguage, native_enum=False), index=True, nullable=False
    )
    short_description: Mapped[str | None] = mapped_column(Text, nullable=True)
    content: Mapped[str] = mapped_column(Text, default="", nullable=False)
    category_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("categories.id"), nullable=True)
    author_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("users.id"), nullable=True)
    status: Mapped[ContentStatus] = mapped_column(
        Enum(ContentStatus, native_enum=False), default=ContentStatus.draft, index=True
    )
    image_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    seo_title: Mapped[str | None] = mapped_column(String(300), nullable=True)
    seo_description: Mapped[str | None] = mapped_column(Text, nullable=True)
    seo_keywords: Mapped[str | None] = mapped_column(String(500), nullable=True)
    published_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    view_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    category = relationship("Category")
    author = relationship("User")
    tags = relationship("Tag", secondary="blog_tags", lazy="selectin")


class BlogTag(Base):
    __tablename__ = "blog_tags"

    blog_id: Mapped[str] = mapped_column(String(36), ForeignKey("blogs.id"), primary_key=True)
    tag_id: Mapped[str] = mapped_column(String(36), ForeignKey("tags.id"), primary_key=True)


class BreakingNews(Base):
    """Ticker headlines shown in the public Breaking News bar."""

    __tablename__ = "breaking_news"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    language: Mapped[ContentLanguage] = mapped_column(
        Enum(ContentLanguage, native_enum=False), index=True, nullable=False
    )
    link_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False, index=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )
